File: ADK_6.2.43/apps/fw/tools/header_path_verify.py
#!/usr/bin/env python2
# Copyright (c) 2016 Qualcomm Technologies International, Ltd.
#   Part of 6.2

"""This script looks for various problems with header files and #includes.

Include directories are any directories begining with "core/include" plus any
hardware-specific subdirectories of hal and io.

Headers that live in an include directory may have common names otherwise they
must be unique.

#includes that reference a file that's not in an include directory must specify
a path.

Example:
    $ cd fw/src && python ../tools/header_path_verify.py
"""

import glob
import os
import re
import sys

"""Files that are allowed to have duplicate names and are allowed to be
included without a path.
"""
INCLUDE_DIR_EXCEPTIONS = (
    "./core/include*/*",
    "./core/hal/*/*/hal/*",
    "./core/io/*/*/io/*",
    "./core/io/*/*/pio_ctrl_sys_io/*",
)

"""Files exempted from the general "unique file names" rule.
There should be a very good reason to use this!!!
"""
DUPLICATE_FILE_EXCEPTIONS = [
    "./customer/core/main/main.h",
]

def _walk(*exts):
    """Walk over each file with an extension in exts."""

    for root, dirs, files in os.walk("."):
        posix_root = root.replace('\\', '/')

        # 'wlan' is from Synergy who have different conventions, so ignore it
        if posix_root in ("./wlan/ip", "./wlan/include"):
            dirs[:] = []
            continue

        for f in files:
            name, ext = os.path.splitext(f)
            if ext in exts:
                yield posix_root + '/' + f

def _include_dir_exceptions():
    return [s.replace('\\', '/')
            for p in INCLUDE_DIR_EXCEPTIONS
            for s in glob.glob(p)]

def _duplicate_name_exceptions():
    return set(_include_dir_exceptions() + DUPLICATE_FILE_EXCEPTIONS)

def duplicate_header_names():
    """Walk over all header files with duplicate names.

    Yields:
        Tuple(str, str): The file that has a duplicated name and the directory
            that the duplicate is in.
    Example:
        >>> from header_path_verify import duplicate_header_names
        >>> print([i for i in duplicate_header_names()])
        [('./path/to/file.c', './duplicate/dir'),
         ('./path/other_file.c', './another/duplicate_dir')]
    """
    visited = {}
    for f in _walk(".h"):
        name = os.path.basename(f)
        if name in visited and f not in _duplicate_name_exceptions():
            yield (f, visited[name])
        visited[name] = os.path.dirname(f)

def _header_filenames_are_unique():
    valid = True
    for f, dup in duplicate_header_names():
        sys.stderr.write("WARNING: found {0} in both {1} and {2}\n".format(
            f, os.path.dirname(f), dup))
        valid = False
    return valid

def includes_without_path():
    """Walk over all includes that do not have a path prefix.

    Walks over all includes that do not have a path prefix. Any headers present
    in the INCLUDE_DIR_EXCEPTIONS list are ignored.

    Yields:
        Tuple(str, int, str): A tuple containing the file name that includes
            a file without a path, the line number that this occurs and the
            header file name that it includes.

    Example:
        >>> from header_path_verify import includes_without_path
        >>> print([i for i in includes_without_path()])
        [('path/to/file.c', 5, 'include.h'),
         ('path/to/other_file.c', 10, 'header.h')]
    """
    files_in_include_dirs = set(os.path.basename(e)
                                for e in _include_dir_exceptions())
    include_re = re.compile("^\s*#\s*include\s+\"([^\"/]+)\"")
    for file_path in _walk(".h", ".c"):
        with open(file_path, 'r') as f:
            for lineno, line in enumerate(f):
                m = include_re.match(line)
                if m:
                    header = m.group(1)
                    if header not in files_in_include_dirs:
                        yield (file_path, lineno, header)

def _includes_contain_path():
    valid = True
    for f, lineno, header in includes_without_path():
        sys.stderr.write("{0}:{1}: WARNING: header {2}:\n"
            "    not in an include directory, included without path\n".format(
                f, str(lineno + 1), header))
        valid = False
    return valid

def validate_includes():
    """Validate include files.

    Traverses the current working dictory and confirms that:
    1. All header files have a unique name, ignoring exceptions in the
       file_exception_strings set.
    2. All includes have a directory prefixed to them, unless the included
       file's directory is in the include_dir_exception_strings set.

    Returns:
        bool: The return value. True if there are no include errors,
            False otherwise.
    Example:
        >>> from header_path_verify import validate_includes
        >>> validate_includes()
        WARNING: found ./private.h in both ./core/windows and ./core/win32
        WARNING: found ./test.h in both ./nfc/nfc_unit and ./nfc/nfc_test
        ./nfc/nfc_unit/llcp_srv_test.c:42: WARNING: header assert.h:
            not in an include directory, included without path
    """
    return all([_header_filenames_are_unique(), _includes_contain_path()])

if __name__ == "__main__":
    sys.exit(not validate_includes())

File: ADK_6.2.43/apps/fw/tools/header_path_verify.pl
#!/usr/bin/perl -w
# Copyright (c) 2016 Qualcomm Technologies International, Ltd.
#   %%version

# This file looks for various problems with headers:
#
# - Headers that don't live in an include directory but are included
#   without specifying the path.  For this purpose, include directories
#   are those beginning with "core/include", plus the hardware-specific
#   subdirectories of hal and io.
#
# - Duplicate names of headers other than those in the include directories,
#   where duplication is allowed to resolve OS, hardware and firmware
#   vs. host variations.
#
# make_all.sh already checks for an explicit "include/", otherwise
# this would do that.
#
# Feel free to rewrite it in Python if you happen to know how to
# do "find" in Python.

use File::Find;
use Data::Dumper;

# Find header files in include directories.
# Also find header files in hal and IO per-chip directories.
our %includes;

my $ret = 0; # Accumulating return code
my $error_count = 0; # For reporting

# Files exempted from the general "unique file names" rule and the directories
# they are alowed to live in. There should be a very good reason to use this!!!
@exemptions = (
    ["hal_macros_divert.h", "hal", "include"],
    ["main.h", "core", "customer"]);
#print Dumper(@exemptions);


foreach my $file (glob("core/include*/*"), glob("core/hal/*/*/hal/*"),
                  glob("core/io/*/*/io/*")) {
    if (-f $file) {
	my $base = $file;
	$base =~ s%^.*/%%;
	$includes{$base} = 1;
    }
}

sub wanted_includes {
    return if $_ eq ".";
    if ($File::Find::dir !~ m%(^|/)include% or
       (-d _ && !m%(^|/)include%)) {
	$File::Find::prune = 1;
	return;
    }

    $includes{$_} = 1  if -f _;
}

find(\&wanted_includes, ".");

# Search through files to see if they include files without a
# / other than those we just found.
#
# Do this recursively --- finds hal and IO stuff.

our %seen;

sub exempt {
    # Some headers are allowed in more than one place.

    $exempt_files_no = $#exemptions + 1;

    $file = shift @_;
    @dirs = @_;
    $dirs_no = $#dirs + 1;

    for ($i = 0; $i < $exempt_files_no; $i++){
        @exemption = @{ $exemptions[$i] };
        #print "@exemption\n";

        $exempt_file = shift @exemption;
        @exempt_dirs = @exemption;

        if ($file eq $exempt_file){
            $missing = 0;
            for ($j = 0; $j < $dirs_no; $j++){
                $dir = $dirs[$j];
                if (not $dir ~~ @exempt_dirs){
                    $missing = 1;
                }
            }
            if (not $missing){
                # File is allowed in the given dirs.
                return 1;
            }
        }
    }

    #File is NOT allowed in the given dirs.
    return 0;
}

sub wanted {
    # $File::Find::dir is the current directory name,
    # $_ is the current filename within that directory
    # $File::Find::name is the complete pathname to the file.
    return unless /\.[ch]$/;

    # Check for header duplicates.
    if (/\.h$/) {
        #print "Considering $_\n";
	if ($seen{$_}) {
	    # Allow duplication within include directories, including
	    # hardware-specific
            # Again, we need the same top_dir stuff
            $top_dir = $File::Find::dir;
            # Get rid of the starting ./
            $top_dir =~ s?./??;
            # Get rid os subdirectories
            $top_dir =~ s?/.*$??;
            # Treat all include directories the same
            $top_dir =~ s?include_.*?include?;
            #print "Comparing $top_dir with $seen{$_}\n";
	    if ($top_dir ne $seen{$_}) {
                if (not exempt($_, $top_dir, $seen{$_})) {
                    print "WARNING: found $_ in both $seen{$_} ",
                    "and $File::Find::dir\n";
                    $ret = 1;
                }
	    }
	} else {
            $top_dir = $File::Find::dir;
            # Get rid of the starting ./
            $top_dir =~ s?./??;
            # Get rid os subdirectories
            $top_dir =~ s?/.*$??;
            # Treat all include directories the same
            $top_dir =~ s?include_.*?include?;
	    $seen{$_} = $top_dir;
            #print "Adding $File: //depot/bg/Releases/ADK/ADK6.2/hydra_os_aura/fw/tools/header_path_verify.pl $_ and dir $top_dir, giving $seen{$_}\n";
	}
    }

    if ($File::Find::dir !~ /^\.\/wlan\/(ip|include)/) {
        if (!open FILE, $_) {
        warn "Can't read $File::Find::name";
        return;
        }

        # Check for headers without paths
        $lineno = 0;
        foreach $line (<FILE>) {
            $lineno++;
        if ($line =~ m%^\s*\#\s*include\s+\"([^\"/]+)\"%) {
            if (!$includes{$1}) {
            print "${File::Find::name}:$lineno: WARNING: header $1:\n",
                        "    not in an include directory, included without path\n";
                    $ret = 1;
                    $error_count += 1;
            }
        }
        }

        close FILE;
    }
}

find(\&wanted, ".");
#print "\nScript $0 found $error_count problems\n";
exit $ret;
